### PR TITLE
MMDiT implementation and text-to-image training with rectified flows

### PR DESCRIPTION
**Pull request by coryMosaicML**
This PR contains an implementation of the MMDiT model from the [SD3 paper](https://arxiv.org/pdf/2403.03206), along with a model class for using it to train text to image models. To support this, a generic model inference class is also included. 

Major additions:
`diffusion/inference/inference_model.py` has `Modelnference` class for inference with arbitrary models from `models.py`
`diffusion/models/models.py` Includes a `text_to_image_transformer` model for SD3 style MMDiT
`diffusion/models/t2i_transformer.py` has the `ComposerModel` class for the MMDiT text to image model
`diffusion/models/transformer.py` has the layers/blocks for the MMDiT model
`diffusion/train.py` includes a new function to configure the optimizer for the new text to image model

### CODE REVIEW

File: diffusion/models/transformer.py [hunk 1]
```diff
@@ -0,0 +1,444 @@
...
+
+    def __init__(self, num_features: int, expansion_factor: int = 4):
+        super().__init__()
+        self.num_features = num_features
+        self.expansion_factor = expansion_factor
+        # AdaLN MLP for post-attention. Initialized to zero so modulation acts as identity at initialization.
+        self.adaLN_mlp_linear = nn.Linear(self.num_features, 4 * self.num_features, bias=True)
+        nn.init.zeros_(self.adaLN_mlp_linear.weight)
+        nn.init.zeros_(self.adaLN_mlp_linear.bias)
+        self.adaLN_mlp = nn.Sequential(nn.SiLU(), self.adaLN_mlp_linear)
```
Comment (A-Jacobson) [inline diffusion/models/transformer.py:203]: probably refactor into something like 

```
AdaptiveLayerNorm(num_features, expansion_factor)
```
same for the one in PreAttentionBlock

File: diffusion/models/t2i_transformer.py [hunk 2]
```diff
@@ -0,0 +1,459 @@
...
+        attention_flops = 4 * self.model.num_layers * seq_len**2 * self.model.num_features * batch_size
+        return 3 * param_flops + 3 * attention_flops
+
+    def encode_image(self, image: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor]:
+        """Encode an image tensor with the autoencoder and patchify the latents."""
+        with torch.cuda.amp.autocast(enabled=False):
+            latents = self.autoencoder.encode(image.half())['latent_dist'].sample().data
+        # Scale and patchify the latents
+        latents = (latents - self.latent_mean) / self.latent_std
+        latent_patches, latent_coords = patchify(latents, self.patch_size)
```
Comment (A-Jacobson) [inline diffusion/models/t2i_transformer.py:215]: I think it makes sense to move the patching/unpatching logic into the `DiffusionTransformer` class. This way the Transformer logic is contained to the transformer and we wouldn't have to pass `coords` to the forward pass. Thoughts?

File: diffusion/models/t2i_transformer.py [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+            return attention_masks
+        elif len(attention_masks.shape) == 3:
+            encoder_attention_masks = attention_masks[:, 0]
+            for i in range(1, attention_masks.shape[1]):
+                encoder_attention_masks |= attention_masks[:, i]
+            return encoder_attention_masks
+        else:
+            raise ValueError(f'attention_mask should have either 2 or 3 dimensions: {attention_masks.shape}')
+
+    def make_text_embeddings_coords(self, text_embeddings: torch.Tensor) -> torch.Tensor:
```
Comment (A-Jacobson) [inline diffusion/models/t2i_transformer.py:255]: Could this also happen in the `DiffusionTransformer` class?

**Review by A-Jacobson — APPROVED** [review]
Left a few commends about structure. Basically, I'd like to see more of the transformer logic confined to the transformer with the `ComposerModel` getting as close to "create the 3 models and call them" as possible. Overall, it's awesome and I conditionally approve as it's non-breaking and also has successful test runs.

File: diffusion/models/t2i_transformer.py (hunk 2 above)
Comment (coryMosaicML) [reply diffusion/models/t2i_transformer.py:215]: I'd like to leave patching/unpatching out of `DiffusionTransformer` as it's specific to doing text to image training without precomputed latents, and in principle `DiffusionTransformer` is agnostic to the modality/preprocessing.

File: diffusion/models/t2i_transformer.py (hunk 3 above)
Comment (coryMosaicML) [reply diffusion/models/t2i_transformer.py:255]: Same as above.

File: diffusion/models/transformer.py (hunk 1 above)
Comment (coryMosaicML) [reply diffusion/models/transformer.py]: Added an `AdaptiveLayerNorm` and also a `ModulationLayer` to replace the duplicated AdaLN logic in the Pre/Post attention blocks.

**Pull request merged**